kernel/comps/virtio/src/queue.rs

```diff
@@ -187,22 +151,182 @@ impl VirtQueue {
+        // pub fn add(
+        //     &mut self,
+        //     inputs: &[&DmaStream],
+        //     outputs: &[&mut DmaStream],
+        // ) -> Result<u16, QueueError> {
+
+        if inputs.is_empty() && outputs.is_empty() {
+            return Err(QueueError::InvalidArgs);
+        }
+        if inputs.len() + outputs.len() > self.desc_table.num_free_desc() {
             return Err(QueueError::BufferTooSmall);
         }
 
-        // allocate descriptors from free list
+        let desc_head = self.desc_table.add_buf(inputs, outputs);
+        self.aval_ring.push_desc(desc_head);
+        Ok(desc_head)
+    }
+
+    /// Returns whether there is an used element that can pop.
+    pub fn can_pop(&self) -> bool {
+        self.used_ring.can_pop()
+    }
+
+    /// Returns the number of free descriptors.
+    pub fn num_free_desc(&self) -> usize {
+        self.desc_table.num_free_desc()
+    }
+
+    /// Pops and returns a token along with the buffer length that
+    /// the device has filled, taken from the used ring buffer.
+    ///
+    /// Ref: linux virtio_ring.c virtqueue_get_buf_ctx
+    pub fn pop_used(&mut self) -> Result<(u16, u32), QueueError> {
+        let Some(used_elem_ptr) = self.used_ring.pop_elem() else {
+            return Err(QueueError::NotReady);
+        };
+
+        let idx = field_ptr!(&used_elem_ptr, UsedElem, idx).read().unwrap();
+        let len = field_ptr!(&used_elem_ptr, UsedElem, len).read().unwrap();
+        self.desc_table.recycle_descriptors(idx as u16);
+        self.used_ring.inc_last_idx();
+        Ok((idx as u16, len))
+    }
+
+    /// If the given token is next on the used ring buffer,
+    /// pops and returns it along with the buffer length that
+    /// the device has filled.
+    ///
+    /// Ref: linux virtio_ring.c virtqueue_get_buf_ctx
+    pub fn pop_used_with_token(&mut self, token: u16) -> Result<u32, QueueError> {
+        let Some(used_elem_ptr) = self.used_ring.pop_elem() else {
+            return Err(QueueError::NotReady);
+        };
+
+        let idx = field_ptr!(&used_elem_ptr, UsedElem, idx).read().unwrap();
+        if idx as u16 != token {
+            return Err(QueueError::WrongToken);
+        }
+        let len = field_ptr!(&used_elem_ptr, UsedElem, len).read().unwrap();
+        self.desc_table.recycle_descriptors(idx as u16);
+        self.used_ring.inc_last_idx();
+        Ok(len)
+    }
+
+    /// Returns the size.
+    pub fn size(&self) -> u16 {
+        SIZE as _
+    }
+
+    /// Returns whether the driver should notify the device.
+    pub fn should_notify(&self) -> bool {
+        self.used_ring.should_notify()
+    }
+
+    /// Notifies the given queue on the device.
+    pub fn notify(&mut self) {
+        self.notify.write(&(self.idx as u32)).unwrap();
+    }
+}
+
+#[derive(Debug)]
+struct DescTable {
+    /// The descriptors
+    descs: Vec<SafePtr<Descriptor, DmaCoherent>>,
```

Comment:
One thing that have annoyed me in the original version is this field of `Vec<_>`. Here is the reasons.

First of all, using `Vec` seems to suggest that the number of descriptors may be adjusted dynamically. But this is not true. From this angle, using `Box<[_]>` is more appropriate. (But as will be seen, using `Box<[_]>` is unnecessary, too)

Second, the `i`-th element of the `Vec` is initialized as a safe pointer that refers to the address of `descriptor_base + i * size_of<Descriptor>`. Storing these safe pointers seems to suggest that they will be updated and cannot be calculated in a straightforward way. But they are NEVER updated. So instead of storing these safe pointers, why not constructing them on the fly?

```rust
#[derive(Debug)]
struct DescTable {
    /// The descriptors
    desc_base: SafePtr<Descriptor, DmaCoherent>,
    ...
}

impl DescTable {
    fn get_desc(&self, idx: usize) -> SafePtr<Descriptor, &DmaCoherent> {
        self.desc_base
            .borrow_vm()
            .add(idx)
    }
}
```

The `get_desc` method above is very light-weight. So we don't have to worry about the overhead of construcinting them on the fly.